File: src/main/java/no/fintlabs/mapping/InstanceReferenceService.java
package no.fintlabs.mapping;

import lombok.*;
import lombok.extern.jackson.Jacksonized;
import lombok.extern.slf4j.Slf4j;
import no.fintlabs.InstanceFieldNotFoundException;
import no.fintlabs.model.instance.InstanceObject;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@Slf4j
@Service
public class InstanceReferenceService {

    private static final Pattern curlyBracketsWrapper = Pattern.compile("\\{[^}]+}");
    private static final Pattern instanceFieldReferencePattern = Pattern.compile("\\$if" + curlyBracketsWrapper);
    private static final Pattern instanceCollectionFieldReferencePattern = Pattern.compile("\\$icf" + curlyBracketsWrapper + curlyBracketsWrapper);
    private static final Pattern referencePattern = Pattern.compile(instanceFieldReferencePattern + "|" + instanceCollectionFieldReferencePattern);

    @Getter
    @Builder(toBuilder = true)
    @Jacksonized
    @EqualsAndHashCode
    private static class CollectionFieldKey {
        private final int collectionIndex;
        private final String collectionFieldKey;
    }

    public String replaceIfReferencesWithInstanceValues(
            String mappingString,
            Map<String, String> instanceValuePerKey,
            InstanceObject[] selectedCollectionObjectsPerKey
    ) {
        Matcher matcher = referencePattern.matcher(mappingString);
        return matcher.replaceAll(matchResult -> {
            String matchedReference = matchResult.group();
            return instanceFieldReferencePattern.matcher(matchedReference).matches()
                    ? getInstanceValue(matchedReference, instanceValuePerKey)
                    : getCollectionFieldValue(matchedReference, selectedCollectionObjectsPerKey);
        });
    }

    private String getInstanceValue(String ifReference, Map<String, String> instanceValuePerKey) {
        String instanceValueKey = getInstanceFieldKey(ifReference);
        if (!instanceValuePerKey.containsKey(instanceValueKey)) {
            throw new InstanceFieldNotFoundException(instanceValueKey);
        }
        return Optional.ofNullable(instanceValuePerKey.get(instanceValueKey)).orElse("");
    }

    private String getInstanceFieldKey(String ifReference) {
        return ifReference.replace("$if{", "").replace("}", "");
    }

    private String getCollectionFieldValue(String icfReference, InstanceObject[] selectedCollectionObjectsPerCollectionIndex) {
        CollectionFieldKey collectionIndexAndFieldReference = getCollectionFieldKey(icfReference);
        Map<String, String> valuePerKeyForCollectionObject = selectedCollectionObjectsPerCollectionIndex[collectionIndexAndFieldReference.getCollectionIndex()].getValuePerKey();
        if (!valuePerKeyForCollectionObject.containsKey(collectionIndexAndFieldReference.getCollectionFieldKey())) {
            throw new InstanceFieldNotFoundException(
                    collectionIndexAndFieldReference.getCollectionFieldKey()
            );
        }
        return Optional.ofNullable(
                valuePerKeyForCollectionObject.get(collectionIndexAndFieldReference.getCollectionFieldKey())
        ).orElse("");
    }

    private CollectionFieldKey getCollectionFieldKey(String icfReference) {
        String[] collectionIndexAndAollectionObjectValueReference = icfReference.split("}\\{");
        int collectionIndex = Integer.parseInt(
                collectionIndexAndAollectionObjectValueReference[0]
                        .replace("$icf{", "")
        );
        String collectionObjectValueReference = collectionIndexAndAollectionObjectValueReference[1]
                .replace("}", "");
        return new CollectionFieldKey(collectionIndex, collectionObjectValueReference);
    }

    public Collection<InstanceObject> getInstanceObjectCollection(
            String collectionReference,
            Map<String, Collection<InstanceObject>> objectCollectionPerKey,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        return instanceFieldReferencePattern.matcher(collectionReference).matches()
                ? getCollectionFromInstance(collectionReference, objectCollectionPerKey)
                : getCollectionFromSelectedCollectionObject(collectionReference, selectedCollectionObjectsByCollectionIndex);
    }

    private Collection<InstanceObject> getCollectionFromInstance(
            String collectionReference,
            Map<String, Collection<InstanceObject>> objectCollectionPerKey
    ) {
        String collectionKey = getInstanceFieldKey(collectionReference);
        if (!objectCollectionPerKey.containsKey(collectionKey)) {
            throw new InstanceFieldNotFoundException(collectionKey);
        }
        return Optional.ofNullable(objectCollectionPerKey.get(collectionKey)).orElse(Collections.emptyList());
    }

    private Collection<InstanceObject> getCollectionFromSelectedCollectionObject(
            String collectionReference,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        CollectionFieldKey collectionFieldKey = getCollectionFieldKey(collectionReference);
        Map<String, Collection<InstanceObject>> objectCollectionPerKey =
                selectedCollectionObjectsByCollectionIndex[collectionFieldKey.getCollectionIndex()]
                        .getObjectCollectionPerKey();
        if (!objectCollectionPerKey.containsKey(collectionFieldKey.getCollectionFieldKey())) {
            throw new InstanceFieldNotFoundException(collectionFieldKey.getCollectionIndex() + "." + collectionFieldKey.getCollectionFieldKey());
        }
        return Optional.ofNullable(objectCollectionPerKey.get(collectionFieldKey.getCollectionFieldKey())).orElse(Collections.emptyList());
    }

}

File: src/main/kotlin/no/novari/flyt/mapping/service/InstanceReferenceService.kt
package no.novari.flyt.mapping.service

import no.novari.flyt.mapping.exception.InstanceFieldNotFoundException
import no.novari.flyt.mapping.model.instance.InstanceObject
import org.springframework.stereotype.Service

@Service
class InstanceReferenceService {
    private data class CollectionFieldKey(
        val collectionIndex: Int,
        val collectionFieldKey: String,
    )

    fun replaceIfReferencesWithInstanceValues(
        mappingString: String,
        instanceValuePerKey: Map<String, String?>,
        selectedCollectionObjectsPerKey: Array<InstanceObject>,
    ): String {
        val matcher = REFERENCE_PATTERN.matcher(mappingString)
        return matcher.replaceAll { matchResult ->
            val matchedReference = matchResult.group()
            if (INSTANCE_FIELD_REFERENCE_PATTERN.matcher(matchedReference).matches()) {
                getInstanceValue(matchedReference, instanceValuePerKey)
            } else {
                getCollectionFieldValue(matchedReference, selectedCollectionObjectsPerKey)
            }
        }
    }

    fun getFirstInstanceValue(
        mappingString: String,
        instanceValuePerKey: Map<String, String?>,
        selectedCollectionObjectsPerKey: Array<InstanceObject>,
    ): String {
        val matcher = REFERENCE_PATTERN.matcher(mappingString)
        if (!matcher.find()) {
            throw IllegalArgumentException("Mapping string contains no valid instance field reference")
        }

        val matchedReference = matcher.group(0)
        return if (INSTANCE_FIELD_REFERENCE_PATTERN.matcher(matchedReference).matches()) {
            getInstanceValue(matchedReference, instanceValuePerKey)
        } else {
            getCollectionFieldValue(matchedReference, selectedCollectionObjectsPerKey)
        }
    }

    fun getInstanceObjectCollection(
        collectionReference: String,
        objectCollectionPerKey: Map<String, Collection<InstanceObject>>,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Collection<InstanceObject> {
        return if (INSTANCE_FIELD_REFERENCE_PATTERN.matcher(collectionReference).matches()) {
            getCollectionFromInstance(collectionReference, objectCollectionPerKey)
        } else {
            getCollectionFromSelectedCollectionObject(collectionReference, selectedCollectionObjectsByCollectionIndex)
        }
    }

    private fun getInstanceValue(
        ifReference: String,
        instanceValuePerKey: Map<String, String?>,
    ): String {
        val instanceValueKey = getInstanceFieldKey(ifReference)
        if (!instanceValuePerKey.containsKey(instanceValueKey)) {
            throw InstanceFieldNotFoundException(instanceValueKey)
        }
        return instanceValuePerKey[instanceValueKey] ?: ""
    }

    private fun getInstanceFieldKey(ifReference: String): String {
        return ifReference
            .replace("${'$'}if{", "")
            .replace("}", "")
    }

    private fun getCollectionFieldValue(
        icfReference: String,
        selectedCollectionObjectsPerCollectionIndex: Array<InstanceObject>,
    ): String {
        val collectionFieldKey = getCollectionFieldKey(icfReference)
        val valuePerKeyForCollectionObject =
            selectedCollectionObjectsPerCollectionIndex[collectionFieldKey.collectionIndex].valuePerKey

        if (!valuePerKeyForCollectionObject.containsKey(collectionFieldKey.collectionFieldKey)) {
            throw InstanceFieldNotFoundException(collectionFieldKey.collectionFieldKey)
        }

        return valuePerKeyForCollectionObject[collectionFieldKey.collectionFieldKey] ?: ""
    }

    private fun getCollectionFieldKey(icfReference: String): CollectionFieldKey {
        val collectionIndexAndCollectionObjectValueReference = icfReference.split("}{")
        val collectionIndex =
            collectionIndexAndCollectionObjectValueReference[0]
                .replace("${'$'}icf{", "")
                .toInt()

        val collectionObjectValueReference =
            collectionIndexAndCollectionObjectValueReference[1]
                .replace("}", "")

        return CollectionFieldKey(
            collectionIndex = collectionIndex,
            collectionFieldKey = collectionObjectValueReference,
        )
    }

    private fun getCollectionFromInstance(
        collectionReference: String,
        objectCollectionPerKey: Map<String, Collection<InstanceObject>>,
    ): Collection<InstanceObject> {
        val collectionKey = getInstanceFieldKey(collectionReference)
        if (!objectCollectionPerKey.containsKey(collectionKey)) {
            throw InstanceFieldNotFoundException(collectionKey)
        }
        return objectCollectionPerKey[collectionKey] ?: emptyList()
    }

    private fun getCollectionFromSelectedCollectionObject(
        collectionReference: String,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Collection<InstanceObject> {
        val collectionFieldKey = getCollectionFieldKey(collectionReference)
        val objectCollectionPerKey =
            selectedCollectionObjectsByCollectionIndex[collectionFieldKey.collectionIndex].objectCollectionPerKey

        if (!objectCollectionPerKey.containsKey(collectionFieldKey.collectionFieldKey)) {
            throw InstanceFieldNotFoundException(
                "${collectionFieldKey.collectionIndex}.${collectionFieldKey.collectionFieldKey}",
            )
        }

        return objectCollectionPerKey[collectionFieldKey.collectionFieldKey] ?: emptyList()
    }

    companion object {
        private const val CURLY_BRACKETS_WRAPPER = """[{][^}]+}"""
        private val INSTANCE_FIELD_REFERENCE_PATTERN =
            Regex("""[$]if$CURLY_BRACKETS_WRAPPER""").toPattern()
        private val INSTANCE_COLLECTION_FIELD_REFERENCE_PATTERN =
            Regex("""[$]icf$CURLY_BRACKETS_WRAPPER$CURLY_BRACKETS_WRAPPER""").toPattern()
        private val REFERENCE_PATTERN =
            Regex(
                "${INSTANCE_FIELD_REFERENCE_PATTERN.pattern()}|${INSTANCE_COLLECTION_FIELD_REFERENCE_PATTERN.pattern()}",
            ).toPattern()
    }
}
